### PR TITLE
Feature/crop isimip3

### PR DESCRIPTION
**Pull request by sameberenz**
_Feature branch updating two modules, hazard.relative_cropyield and exposures.crop_production._

**Main changes:**
- Rename key methods and functions in both modules to make their functionality more clear and identify methods and functions that are specific to input data from the ISMIP project more clearly. Renamings:

   1. relative_cropyield (hazard):
      - set_from_single_run() --> set_from_isimip_netcdf()
      - generate_full_hazard_set() --> set_multiple_rc_from_isimip()
      - init_hazard_set() --> init_hazard_sets_isimip()
      - calc_his_haz() --> calc_his_haz_isimip()
      - calc_fut_haz() --> calc_fut_haz_isimip()
   2. crop_production (exposure):
      - set_from_single_run() --> set_from_isimip_netcdf()
      - init_full_exposure_set() --> init_full_exp_set_isimip()
'  
- Make both modules compatible with output data from **ISIMIP3b** (so far: 2a and 2b).
- CropProduction(): add method **set_to_kcal()** to convert unit from t / y to kcal / y.
- Clean up code and doc strings.

### CODE REVIEW

File: climada/entity/exposures/crop_production.py [hunk 1]
```diff
@@ -142,19 +155,38 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
...
-
+        if not bbox:
+            bbox = BBOX
+        if not input_dir:
+            input_dir = INPUT_DIR
+        if hist_mean is None:
+            hist_mean = HIST_MEAN_PATH
+        if not fn_str_var:
+            fn_str_var = FN_STR_VAR
+        if (not isimip_version) or ('ISIMIP2' in isimip_version):
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:175]: That is strange. Why do you redefine `isimip_version` when `'ISIMIP2` is in `isimip_version?

File: climada/entity/exposures/crop_production.py [hunk 2]
```diff
@@ -142,19 +155,38 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
...
+            isimip_version = 'ISIMIP2'
+        elif 'ISIMIP3' in isimip_version:
+            isimip_version = 'ISIMIP3'
+        if (not scenario) or ('hist' in scenario):
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:179]: Why this strange or clause here?

File: climada/entity/exposures/crop_production.py [hunk 3]
```diff
@@ -218,7 +250,20 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
 
         # set historic mean, its latitude, and longitude:
         hist_mean_dict = dict()
-        if isdir(hist_mean):
+        # if hist_mean is given as np.ndarray or dict,
+        # code assumes it contains hist_mean as returned by the hazard crop_potential
+        # however structured in dictionary as hist_mean_dict, with same
+        # bbox extensions as the exposure:
+        if isinstance(hist_mean, dict):
+            if not ('firr' in hist_mean.keys() or 'noirr' in hist_mean.keys()):
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:258]: What about `combined`?

File: climada/entity/exposures/crop_production.py [hunk 4]
```diff
@@ -218,7 +250,20 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
...
+                LOGGER.error('Invalid hist_mean provided: {hist_mean}')
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:259]: the Logger prefers to be lazy.

```suggestion
                LOGGER.error('Invalid hist_mean provided: %s' %hist_mean)
```

File: climada/entity/exposures/crop_production.py [hunk 5]
```diff
@@ -218,7 +250,20 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
...
+                raise ValueError('invalid hist_mean.')
+            hist_mean_dict = hist_mean
+            lat_mean = self.latitude.values
+        elif isinstance(hist_mean, np.ndarray):
+            hist_mean_dict[irr[0]] = hist_mean
+            lat_mean = self.latitude.values
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:265]: This is unclear. Why is this needed? What happens if `hist_mean` is a list?

File: climada/entity/exposures/crop_production.py [hunk 6]
```diff
@@ -238,9 +283,8 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
             lat_mean = hist_mean['lat'][()]
             lon_mean = hist_mean['lon'][()]
         else:
-        # hist_mean as returned by the hazard crop_potential is used (array format) with same
-        # bbox extensions as the exposure
-            lat_mean = self.latitude.values
+            LOGGER.error('Invalid hist_mean provided: {hist_mean}')
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:286]:
```suggestion
            LOGGER.error('Invalid hist_mean provided: %s ' %hist_mean')
```

File: climada/entity/exposures/crop_production.py [hunk 7]
```diff
@@ -256,8 +300,11 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
         # The exposure [t/y] is computed per grid cell as the product of the area covered
         # by a crop [ha] and its yield [t/ha/y]
         self['value'] = np.squeeze(area_crop[irr[0]]*hist_mean_dict[irr[0]][idx_mean])
+        self['value'] = np.nan_to_num(self.value) # replace NaN by 0.0
         for irr_val in irr[1:]: # add other irrigation types if irr=combined
-            self['value'] += np.squeeze(area_crop[irr_val]*hist_mean_dict[irr_val][idx_mean])
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:260]: That is not a very clear use. Why would this be needed iif `irr = combined`? What happens if e.g. `irr = ['firr', 'combined', None]` ?

File: climada/entity/exposures/crop_production.py [hunk 8]
```diff
@@ -286,18 +333,22 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
...
 
         return self
 
-    def set_mean_of_several_models(self, input_dir=INPUT_DIR, hist_mean=HIST_MEAN_PATH, bbox=BBOX,
-                                   yearrange=(YEARCHUNKS['histsoc'])['yearrange'],
-                                   cl_model=None, scenario=None, crop=None, irr=None,
-                                   unit='USD', fn_str_var=FN_STR_VAR):
-        """Wrapper to fill exposure from several nc_dis files from ISIMIP
+    def set_mean_of_several_models(self, input_dir=None, hist_mean=None, bbox=None,
+                                   yearrange=None, cl_model=None, scenario=None,
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:347]: Several `Isimip` models?

**Review by chahank — CHANGES_REQUESTED** [review]
Looks good overall, small questions/suggestions. I am about 1/2 through.

File: climada/entity/exposures/crop_production.py [hunk 9]
```diff
@@ -366,7 +434,20 @@ def set_mean_of_several_models(self, input_dir=INPUT_DIR, hist_mean=HIST_MEAN_PA
...
-    def set_to_usd(self, input_dir=INPUT_DIR, yearrange=YEARS_FAO):
+    def set_to_kcal(self):
+        """Converts the exposure from tonnes to kcal using conversion factor per crop type.
+
+        Returns:
+            Exposure
+        """
+        if not 't' in self.value_unit:
+            LOGGER.warning('self.unit is neither t nor t / year.')
+        self['tonnes_per_year'] = self['value'].values
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:445]: How can the unit be 't' but the value is `tonnes_per_year` ?

File: climada/entity/exposures/crop_production.py [hunk 10]
```diff
@@ -366,7 +434,20 @@ def set_mean_of_several_models(self, input_dir=INPUT_DIR, hist_mean=HIST_MEAN_PA
...
         return self
 
-    def set_to_usd(self, input_dir=INPUT_DIR, yearrange=YEARS_FAO):
+    def set_to_kcal(self):
+        """Converts the exposure from tonnes to kcal using conversion factor per crop type.
+
+        Returns:
+            Exposure
+        """
+        if not 't' in self.value_unit:
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:443]: What if a unit which contains the letter `t` but is different from `t` or `t/y` is set by a user?

File: climada/entity/exposures/crop_production.py [hunk 11]
```diff
@@ -366,7 +434,20 @@ def set_mean_of_several_models(self, input_dir=INPUT_DIR, hist_mean=HIST_MEAN_PA
...
+            LOGGER.warning('self.unit is neither t nor t / year.')
+        self['tonnes_per_year'] = self['value'].values
+        self.value = self.value * KCAL_PER_TON[self.crop]
+        self.value_unit = 'kcal / y'
+        return self
+
+    def set_to_usd(self, input_dir=None, yearrange=None):
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:450]: Set what to usd?

File: climada/entity/exposures/crop_production.py [hunk 12]
```diff
@@ -499,24 +597,25 @@ def init_full_exposure_set(input_dir=INPUT_DIR, filename=None, hist_mean_dir=HIS
...
     return filename_list, output_list
 
-def normalize_with_fao_cp(exp_firr, exp_noirr, input_dir=INPUT_DIR,
-                          yearrange=np.array([2008, 2018]), unit='t', return_data=True):
-    """Normalize the given exposures countrywise with the mean crop production quantity
-    documented by the FAO. Refer to the beginning of the script for guidance on where to
-    download the needed FAO data.
+def normalize_with_fao_cp(exp_firr, exp_noirr, input_dir=None,
+                          yearrange=None, unit=None, return_data=True):
+    """Normalize (i.e., bias corrent) the given exposures countrywise with the mean
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:615]: corrent = ?

File: climada/entity/exposures/crop_production.py [hunk 13]
```diff
@@ -546,14 +645,19 @@ def normalize_with_fao_cp(exp_firr, exp_noirr, input_dir=INPUT_DIR,
         exp_tot_production(list): Exposure crop production value per country
             (before normalization)
     """
-
-    # if the exposure unit is USD/y temporarily reset the exposure to t/y
+    if not input_dir:
+        input_dir = INPUT_DIR
+    if yearrange is None:
+        yearrange = (2008, 2018)
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:651]: why not the `YEARS_FAO`

File: climada/entity/exposures/crop_production.py [hunk 14]
```diff
@@ -546,14 +645,19 @@ def normalize_with_fao_cp(exp_firr, exp_noirr, input_dir=INPUT_DIR,
...
+    # if the exposure unit is USD/y or kcal/y, temporarily reset the exposure to t/y
     # (stored in tonnes_per_year) in order to normalize with FAO crop production
-    # values and then apply set_to_USD() for the normalized exposure to restore the
+    # values and then apply set_to_XXX() for the normalized exposure to restore the
     # initial exposure unit
-    if exp_firr.value_unit == 'USD / y':
+    if exp_firr.value_unit == 'USD / y' or 'kcal' in exp_firr.value_unit:
         exp_firr.value = exp_firr.tonnes_per_year
-    if exp_noirr.value_unit == 'USD / y':
+    if exp_noirr.value_unit == 'USD / y' or 'kcal' in exp_noirr.value_unit:
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:660]: Why would the value but set to `tonnes_per_year` if the unit is `USD / y` ?

File: climada/entity/exposures/crop_production.py [hunk 15]
```diff
@@ -638,6 +747,14 @@ def normalize_several_exp(input_dir=INPUT_DIR, output_dir=OUTPUT_DIR,
...
                 (before normalization)
     """
+    if not input_dir:
+        input_dir = INPUT_DIR
+    if not output_dir:
+        output_dir = OUTPUT_DIR
+    if not unit:
+        unit = 't'
+    if yearrange is None:
+        yearrange = (2008, 2018)
```
Comment (chahank) [inline climada/entity/exposures/crop_production.py:757]: Why `(2008, 2018)`? Default constants should be set in constants or in the config file.

File: climada/hazard/relative_cropyield.py [hunk 16]
```diff
@@ -596,89 +635,113 @@ def init_hazard_set(filenames, input_dir=INPUT_DIR, bbox=BBOX, isimip_run='ISIMI
...
+                                              file_props[his_file_list[0]]['endyear']))
 
     # initiate the historic mean for each combination of crop and irrigation type
     # the idx keeps track of the row in which the hist_mean values are written per crop-irr to
     # ensure that all files are assigned to the corresponding crop-irr combination
     hist_mean_per_crop = dict()
-    for crop_irr in crop_list:
-        crop, irr = crop_irr.split('-')
+    for i, combi_crop_irr in enumerate(combi_crop_list):
+        crop, irr = crop_list[i].split('-')
```
Comment (chahank) [inline climada/hazard/relative_cropyield.py:731]: Why does `crop_list` contain something else than crops?

```suggestion
    for crops, combi_crop_irr in zip(crop_list, combi_crop_list):
        crop, irr = crops.split('-')
```

**Review by chahank — CHANGES_REQUESTED** [review]

File: climada/entity/exposures/crop_production.py (hunk 1 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:175]: This is to allow for inputs for isimip_version like 'ISIMIP2a' instead of 'ISIMIP2'. If input is 'ISIMIP2a', it is automatically reduced to 'ISIMIP2'. I would like to keep that tolerance in for convenience of users, because in the module relative_cropyield, the differenciation between **a** and **b** plays a role, while in crop_production it doesn't....

How about if I change it to the following to be less confusing, @chahank?

`if (not isimip_version) or (isimip_version in ['ISIMIP2a, 'ISIMIP2b']):`

File: climada/entity/exposures/crop_production.py (hunk 2 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:179]: their are different variants of scenario's that all require the same choice here, how about:

` if (not scenario) or (scenario in ('historical', 'hist')):`

File: climada/entity/exposures/crop_production.py (hunk 3 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:258]: hist_mean is never `combined`, it always provides either 'firr' or 'noirr' data. If users choose `irr='combined'`, the hist_mean is loaded for each irrigation type and combined further below:

In lines 238f, the input irr='combined' is translated into the two combination types:
```
        if irr == 'combined':
            irr = ['firr', 'noirr']
```

in lines 304f, they are combined:
```
        for irr_val in irr[1:]: # add other irrigation types if irr=combined
            value_tmp = np.squeeze(area_crop[irr_val]*hist_mean_dict[irr_val][idx_mean])
            value_tmp = np.nan_to_num(value_tmp) # replace NaN by 0.0
            self['value'] += value_tmp
```

should this be explained better with comments?

File: climada/entity/exposures/crop_production.py (hunk 5 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:265]: hist_mean can be given in different variable types. if only one irrigation type is considered, it is usually an array. I can add the option list here as well

File: climada/entity/exposures/crop_production.py (hunk 7 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:260]: There are three options for irr:
_firr_, _noirr_, or _combined_. Only if `irr=='combined'`, the variable is changed to the list `['firr', 'noirr']`.

`irr = ['firr', 'combined', None]`  is not a valid option, if users provide this as input there will be a ValueError because there is no `area_crop['combined']`

File: climada/entity/exposures/crop_production.py [hunk 17]
```diff
@@ -354,9 +422,9 @@ def set_mean_of_several_models(self, input_dir=INPUT_DIR, hist_mean=HIST_MEAN_PA
         # The calculations are repeated for all remaining exposures (starting from index 1 as
         # the first exposure has been saved in combined_exp[:, 0])
         for j in range(1, len(filenames['subset'])):
```
[the remark this answers is not in the document]
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:424]: The variable **j** is required below in the loop, but I will change it to use `enumerate`

File: climada/entity/exposures/crop_production.py (hunk 9 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:445]: this will be adjusted

File: climada/entity/exposures/crop_production.py (hunk 11 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:450]: how about `set_value_to_usd()` ?

File: climada/entity/exposures/crop_production.py (hunk 12 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:615]: "correct"

File: climada/entity/exposures/crop_production.py (hunk 10 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:443]: There is some confusion between "tonnes" and "t" and "t / y", ... I will clean this up.
However, if the user sets a unit manually, that's probably intentional and assumed to be for a reason, can't catch that.

File: climada/entity/exposures/crop_production.py (hunk 14 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:660]: we had this discussion in the last code review already, here is the explanatory comment we added as a consequence:

    # if the exposure unit is USD/y or kcal/y, temporarily reset the exposure to t/y
    # (stored in tonnes_per_year) in order to normalize with FAO crop production
    # values and then apply set_to_XXX() for the normalized exposure to restore the
    # initial exposure unit

File: climada/hazard/relative_cropyield.py (hunk 16 above)
Comment (sameberenz) [reply climada/hazard/relative_cropyield.py:731]: The variable names are confusing here, because the term "combination" is used for two different kinds of combinations...
combined crops are for instance ri1 and ri2 (two seasons of rice) combined to "ric", while a crop-irr combination is for instance "ric-firr"
```
            crop_list (list): list of all crop-irr combinations
            combi_crop_list (list): list of all crop-irr combinations for combined crops
```
how about renaming the variables as following?
```
            crop_irr_list (list): list of all crop-irr combinations
            combi_crop_irr_list (list): list of all crop-irr combinations for combined crops
```

**Comment by sameberenz** [comment]
I've commited and pushed proposed changes in repsonse to all comments markes as "resolved".
There are 4 comments still unresolved where I'd like your input @chahank 
thanks for your code review and support!

File: climada/hazard/relative_cropyield.py [hunk 18]
```diff
@@ -690,44 +753,74 @@ def calc_his_haz(his_file, file_props, input_dir=INPUT_DIR, bbox=BBOX, yearrange
...
+        else:
+            his_file2 = None
+        if his_file2 and isfile(join(input_dir, his_file2)):
+            haz_his2 = RelativeCropyield()
+            haz_his2.set_from_isimip_netcdf(input_dir=input_dir, filename=his_file2, bbox=bbox,
+                                scenario=file_props[his_file]['scenario'],
+                                yearrange=np.array([file_props[his_file]['startyear'],
+                                                    file_props[his_file]['endyear']]))
+            if crop == 'swh':
+            # mask of winter wheat in spring wheat and vice versa:
```
Comment (carmensteinmann) [inline climada/hazard/relative_cropyield.py:785]: I am not sure whether I understand what the mask does. What does the file winter_and_spring_wheat_areas_phase3.nc4 contain?

File: climada/hazard/relative_cropyield.py [hunk 19]
```diff
@@ -444,34 +451,38 @@ def plot_comparing_maps(self, his, fut, axes, nr_cli_models=1, model=1):
...
-            isimip_run (string): name of the ISIMIP run (ISIMIP2a or ISIMIP2b)
+            isimip_run (string): name of the ISIMIP run (f.i. ISIMIP2a or ISIMIP2b)
             yearrange_his (int tuple): year range for the historical hazard sets
             yearrange_mean (int tuple): year range for the historical mean
             return_data (boolean): returned output
                 False: returns list of filenames only
                 True: returns also list of data
             save (boolean): save output data to output_dir
-
+            combine_subcrops (boolean): combine coprs: ric=ri1+ri2, whe=swh+wwh
```
Comment (carmensteinmann) [inline climada/hazard/relative_cropyield.py:473]:
```suggestion
            combine_subcrops (boolean): combine crops: ric=ri1+ri2, whe=swh+wwh
```

File: climada/hazard/relative_cropyield.py [hunk 20]
```diff
@@ -743,35 +836,105 @@ def calc_fut_haz(his_file, scenario, file_props, hist_mean, input_dir=INPUT_DIR,
...
             haz_fut (RelativeCropyield): future hazard
-            filename (string): name to save historical hazard
+            filename (string): name to save future hazard
+    """
+    if yearrange_fut is None:
+        if isinstance(fut_file, str) and (len(fut_file.split('_')[-2]) == 4):
+            yearrange_fut = (int(fut_file.split('_')[-2]),
+                             int(fut_file.split('_')[-1].split('.')[0])
+                             )
+        else: # define yearrane from defaults if not specified by user or filename:
```
Comment (carmensteinmann) [inline climada/hazard/relative_cropyield.py:851]:
```suggestion
        else: # define yearrange from defaults if not specified by user or filename:
```

**Review by carmensteinmann — CHANGES_REQUESTED** [review]

File: climada/entity/exposures/crop_production.py (hunk 3 above)
Comment (chahank) [reply climada/entity/exposures/crop_production.py:258]: You know that I am always in favor of good comment :D

File: climada/entity/exposures/crop_production.py (hunk 7 above)
Comment (chahank) [reply climada/entity/exposures/crop_production.py:260]: That is suboptimal. A variable should not change type during the code. It should not go from str to list.

File: climada/hazard/relative_cropyield.py (hunk 18 above)
Comment (sameberenz) [reply climada/hazard/relative_cropyield.py:785]: The masks in the NetCDF file divides all wheat growing areas globally in two distinct categories, either growing winter wheat (wwh) or spring wheat (swh). Since the hazard sets for wwh and swh are calculated for the whole globel ("all crops everywhere") we need to decide for each grid cell which one to take when combining wwh and swh into one combined crop wheat (whe). 

I will add this explanation as a comment in the script.

File: climada/entity/exposures/crop_production.py (hunk 14 above)
Comment (chahank) [reply climada/entity/exposures/crop_production.py]: Aaa yes. I am still not very happy with this, but ok.

File: climada/hazard/relative_cropyield.py (hunk 16 above)
Comment (chahank) [reply climada/hazard/relative_cropyield.py:731]: good idea.

File: climada/entity/exposures/crop_production.py (hunk 7 above)
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:260]: agreed, I will disentangle this by defining an extra list variable `irr_types`

File: climada/entity/exposures/crop_production.py [hunk 21]
```diff
@@ -204,7 +236,7 @@ def set_from_single_run(self, input_dir=INPUT_DIR, filename=None, hist_mean=HIST
         # The area covered by a crop is calculated as the product of the fraction and
         # the grid cell size
         if irr == 'combined':
-            irr = ['irr', 'noirr']
+            irr = ['firr', 'noirr']
         else:
             irr = [irr]
```
[the remark this answers is not in the document]
Comment (sameberenz) [reply climada/entity/exposures/crop_production.py:241]: I'll add a check setting it to 'combined' if irr is None or not provided

**Comment by sameberenz** [comment]
@chahank @carmensteinmann , I have now answered to all comments and commited changes to the code as suggested. Could you please briefly check and approve the pull request if satisfied? Thanks

**Review by carmensteinmann — APPROVED** [review]
